astropy/nddata/nddata.py

```diff
@@ -198,17 +213,23 @@ def __init__(
                 else:
                     mask = data.mask
 
-                # Just save the data for further processing, we could be given
-                # a masked Quantity or something else entirely. Better to check
-                # it first.
-                data = data.data
-
-            if isinstance(data, Quantity):
-                if unit is not None and unit != data.unit:
-                    log.info("overwriting Quantity's current unit with specified unit.")
-                else:
-                    unit = data.unit
-                data = data.value
+                # get the (possibly masked quantity) data from CCDData
+                # data = data.data
+
+            # if data inherits from ndarray, continue:
+            mro = data.__class__.__mro__
+            if np.ndarray in mro:
+                # if data is a MaskedQuantity, then the first base
+                # is the quantity, second is the numpy ndarray. The
+                # third would be a flat ndarray, but we don't want to
+                # go that deep, so stop getting bases:
+                while np.ndarray not in mro[:2]:
+                    if data.base is not None:
+                        data = data.base
+                    if issubclass(data.__class__, Quantity):
+                        unit = data.unit
+                        data = data.value
+                    mro = data.__class__.__mro__
```

Comment:
This block handles masked quantities and other Masked `np.ndarray` subclasses. Suppose you initialize `NDDataArray` like so
```python
import numpy as np
import astropy.units as u
from astropy.utils.masked import Masked

a = np.array([1, 2, 3, 4, 5])
mask = [True, False, False, True, True]
mqa = Masked(u.Quantity(a, u.m), mask=mask)
nd1 = NDDataArray(mqa)
```
then `NDDataArray` should get the values, mask, and unit from `mqa`. This turns out to be a bit involved, because `type(mqa)` is `astropy.utils.masked.core.MaskedQuantity`, and we have to use the `Masked` API to get the data values. To get the underlying quantity without a mask you can do `mqa.base`, which returns `[1, 2, 3, 4, 5] m`. To get the underlying data values in that quantity you can do `mqa.base.base` or `mqa.base.value`, which both return `array([1., 2., 3., 4., 5.])`. 

The highlighted block allows us to repeatedly go up the inheritance tree of `data` by taking `data.base` until the underlying `ndarray` and its associated masks and units are extracted.